src/writer/libtest.rs

```diff
@@ -0,0 +1,708 @@
+
+use std::{
+    fmt::Debug,
+    io, iter, mem,
+    str::FromStr,
+    time::{Duration, SystemTime},
+};
+
+use async_trait::async_trait;
+use derive_more::From;
+use either::Either;
+use itertools::Itertools as _;
+use serde::Serialize;
+
+use crate::{
+    event, parser,
+    writer::{self, basic::coerce_error, out::WriteStrExt as _, Arbitrary},
+    Event, World, Writer,
+};
+
+/// CLI options of a [`Libtest`] [`Writer`].
+#[derive(Debug, Clone, clap::Args)]
+pub struct Cli {
+    /// Configure formatting of output.
+    #[clap(long, name = "json")]
+    pub format: Option<Format>,
+
+    /// Show captured stdout of successful tests. Currently does nothing.
+    #[clap(long)]
+    pub show_output: bool,
+
+    /// Enable nightly-only flags.
+    #[clap(short = 'Z')]
+    pub nightly: Option<String>,
+}
+
+/// Output formats.
+///
+/// Currently supports only `JSON`.
+#[derive(Clone, Copy, Debug)]
+pub enum Format {
+    /// [`libtest`][1]'s `JSON` format.
+    ///
+    /// [1]: https://doc.rust-lang.org/rustc/tests/index.html
+    Json,
+}
+
+impl FromStr for Format {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_str() {
+            "json" => Ok(Self::Json),
+            s @ ("pretty" | "terse" | "junit") => {
+                Err(format!("`{s}` option is not supported yet"))
+            }
+            s => Err(format!(
+                "Unknown option `{s}`, expected `pretty` or `json`",
+            )),
+        }
+    }
+}
+
+/// [`libtest`][1] compatible [`Writer`].
+///
+/// Currently used only to support `--format=json` option.
+///
+/// # Ordering
+///
+/// This [`Writer`] isn't [`Normalized`] by itself, so should be wrapped into
+/// a [`writer::Normalize`], otherwise will produce output [`Event`]s in a
+/// broken order.
+/// Ideally, we shouldn't wrap this into [`writer::Normalize`] and leave this to
+/// tools, parsing JSON output. Unfortunately, not all tools can do that (ex.
+/// [`IntelliJ Rust`][2]), so it's still recommended to wrap this into
+/// [`writer::Normalize`] even if it can mess up timing reports.
+///
+/// [1]: https://doc.rust-lang.org/rustc/tests/index.html
+/// [2]: https://github.com/intellij-rust/intellij-rust/issues/9041
+/// [`Normalized`]: writer::Normalized
+#[derive(Clone, Debug)]
+pub struct Libtest<W, Out: io::Write = io::Stdout> {
+    /// [`io::Write`] implementor to output into.
+    output: Out,
+
+    /// Collection of events before [`ParsingFinished`] is received.
+    ///
+    /// Until [`ParsingFinished`] is received, all events are stored inside
+    /// [`Libtest::events`] and outputted only after that event is received.
+    /// This is done, because [`libtest`][1]'s first event must contain number
+    /// of executed test cases.
+    ///
+    /// [1]: https://doc.rust-lang.org/rustc/tests/index.html
+    /// [`ParsingFinished`]: event::Cucumber::ParsingFinished
```

Comment:
@ilslv it would be nice to have any link leading to format description. It's quite hard to find one by hand.

Reply:
@ilslv well... at least a comment, or a source file. It should be defined somewhere and anyhow.